### PR TITLE
Implement `DataFrame.to_parquet`

### PR DESCRIPTION
**Pull request by rjzamora**
Note that I experimented with a few variations on this, and ultimately decided to keep the `Expr` component of `to_parquet` as simple as possible.

### CODE REVIEW

File: dask_expr/io/parquet.py [hunk 1]
```diff
@@ -62,6 +68,309 @@ def normalize_pa_schema(schema):
...
+                        self.path,
+                    ],
+                    {"append": append, "compression": compression},
+                )
+            }
+        else:
+            return {(self._name, 0): (lambda x: None, self.frame.__dask_keys__())}
+
+
+def to_parquet(
```
Comment (rjzamora) [inline dask_expr/io/parquet.py:155]: NOTE: This function is mostly copied from `dd.DataFrame.to_parquet`. I didn't find any benefit in moving any option validation of filesystem/overwriting logic into an `Expr` class.

File: dask_expr/io/parquet.py [hunk 2]
```diff
@@ -62,6 +68,309 @@ def normalize_pa_schema(schema):
...
+                engine,
+                i_offset,
+                partition_on,
+                write_metadata_file,
+                name_function,
+                toolz.merge(
+                    kwargs,
+                    {"compression": compression, "custom_metadata": custom_metadata},
+                    extra_write_kwargs,
+                ),
```
Comment (rjzamora) [inline dask_expr/io/parquet.py:359]: Only change to `to_parquet` is here.

File: dask_expr/io/parquet.py [hunk 3]
```diff
@@ -62,6 +68,309 @@ def normalize_pa_schema(schema):
...
+            # output directory, since deleting those files now would result in
+            # errors or incorrect results.
+            for read_op in df.expr.find_operations(ReadParquet):
+                read_path_with_slash = str(read_op.path).rstrip("/") + "/"
+                write_path_with_slash = path.rstrip("/") + "/"
+                if read_path_with_slash.startswith(write_path_with_slash):
+                    raise ValueError(
+                        "Cannot overwrite a path that you are reading "
+                        "from in the same task graph."
+                    )
```
Comment (rjzamora) [inline dask_expr/io/parquet.py:225]: and here.

File: dask_expr/io/parquet.py [hunk 4]
```diff
@@ -62,6 +68,309 @@ def normalize_pa_schema(schema):
...
+class ToParquetBarrier(Expr):
+    _parameters = ToParquet._parameters
+
+    @property
+    def _meta(self):
+        return None
+
+    def _divisions(self):
+        return (None, None)
+
```
Comment (mrocklin) [inline dask_expr/io/parquet.py:133]: Maybe we need to be ok with `Expr`s that aren't dataframes.  

Nothing to do at the moment, but this reinforces a questions we've had for a while.  Just pointing it out here.

File: dask_expr/collection.py [hunk 5]
```diff
@@ -573,6 +573,11 @@ def dropna(self, how=no_default, subset=None, thresh=no_default):
...
         return new_collection(RenameFrame(self.expr, columns=columns))
 
+    def to_parquet(self, path, **kwargs):
+        from dask_expr.io.parquet import to_parquet
+
+        return to_parquet(self, path, **kwargs)
+
 
 class Series(FrameBase):
     """Series-like Expr Collection"""
```
Comment (mrocklin) [inline dask_expr/collection.py:583]: Is the implementation above only on DataFrames?  If so do we need a Series implementation (which I assume is just `to_frame().to_parquet(...)`

**Review by mrocklin — APPROVED** [review]
Broadly it seems fine to me.  Thanks for doing this @rjzamora

File: dask_expr/collection.py (hunk 5 above)
Comment (mrocklin) [reply dask_expr/collection.py:583]: Ah, ok, that makes sense

**Pull request merged**